driver-core/src/main/com/mongodb/operation/OperationHelper.java

```diff
@@ -414,6 +420,35 @@ static boolean serverIsAtLeastVersion(final ConnectionDescription description, f
         }
     }
 
+    static <T> T withReleasableConnection(final WriteBinding binding, final CallableWithConnectionAndSource<T> callable) {
+        ConnectionSource source = binding.getWriteConnectionSource();
+        try {
+            return callable.call(source, source.getConnection());
+        } finally {
+            source.release();
+        }
+    }
+
+    static <T> T withReleasableConnection(final WriteBinding binding, final Throwable connectionException,
+                                          final CallableWithConnectionAndSource<T> callable) {
+        ConnectionSource source = null;
+        Connection connection = null;
```

Comment:
Nit: According to IntelliJ an extraneous assignment to null

Reply:
Done